### PR TITLE
S3: log human readable error on connection failure

### PR DESCRIPTION
**Pull request by earl-warren**
Should BucketExists (HeadBucket) fail because of an error related to
the connection rather than the existence of the bucket, no information
is available and the admin is left guessing.

https://docs.aws.amazon.com/AmazonS3/latest/API/API_HeadBucket.html

> This action is useful to determine if a bucket exists and you have
> permission to access it. The action returns a 200 OK if the bucket
> exists and you have permission to access it.
>
> If the bucket does not exist or you do not have permission to access
> it, the HEAD request returns a generic 400 Bad Request, 403
> Forbidden or 404 Not Found code. A message body is not included, so
> you cannot determine the exception beyond these error codes.

GetBucketVersioning is used instead and exclusively dedicated to
asserting if using the connection does not return a BadRequest.
If it does the NewMinioStorage logs an error and returns. Otherwise
it keeps going knowing that BucketExists is not going to fail for
reasons unrelated to the existence of the bucket and the permissions
to access it.

(cherry picked from commit d1df4b3bc62e5e61893a923f1c4b58f084eb03af)

Refs: https://codeberg.org/forgejo/forgejo/issues/1338

### CODE REVIEW

**Comment by puni9869** [comment]
Some linter errors.

**Comment by wxiaoguang** [comment]
Instead of introducing `mockvariable`, actually it should do this:

->  Move web/api context related testing function into a separate package #26859 


See the TODO comment of the MockContext:

```
// TODO: move this function to other packages, because it depends on "models" package
```

**Comment by earl-warren** [comment]
The commit related to MockVariableValue was removed, it does not belong here.

File: modules/storage/minio.go [hunk 1]
```diff
@@ -90,6 +95,19 @@ func NewMinioStorage(ctx context.Context, cfg *setting.Storage) (ObjectStorage,
 		return nil, convertMinioErr(err)
 	}
 
+	// Check if the connection works
+	err = getBucketVersioning(ctx, minioClient, config.Bucket)
```
Comment (lunny) [inline modules/storage/minio.go:99]: Why not invoke `_, err := minioClient.GetBucketVersioning(ctx, config.Bucket)` directly here?

File: modules/storage/minio.go [hunk 2]
```diff
@@ -90,6 +95,19 @@ func NewMinioStorage(ctx context.Context, cfg *setting.Storage) (ObjectStorage,
...
+	if err != nil {
+		errResp, ok := err.(minio.ErrorResponse)
+		if !ok {
+			return nil, err
+		}
+		if errResp.StatusCode == http.StatusBadRequest {
```
Comment (lunny) [inline modules/storage/minio.go:109]: Why not handle other status codes, such as `StatusNotFound`, `StatusForbidden`, `StatusConflict` and `StatusPreconditionFailed`?

Comment (earl-warren) [reply modules/storage/minio.go:109]: It is explained in the PR description.

**Comment by wxiaoguang** [comment]
I am not sure whether it is right. According to https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetBucketVersioning.html

> Returns the versioning state of a bucket. To retrieve the versioning state of a bucket, you must be the bucket owner.

Then, are you sure that the non-owner but writer users could still run Gitea correctly?

**Comment by earl-warren** [comment]
> GetBucketVersioning is used instead and exclusively dedicated to asserting if using the connection does not return a BadRequest.

**Comment by wxiaoguang** [comment]
> > GetBucketVersioning is used instead and exclusively dedicated to asserting if using the connection does not return a BadRequest.

That's quite tricky. Could there be some comments to explain the design and behavior? eg:

```
// The GetBucketVersioning is only used for checking whether the Object Storage parameters are generally good. It doesn't need to succeed.
// The assumption is that if the API returns the HTTP code 400, then the parameters could be incorrect. Otherwise even if the request itself fails (403, 404, etc), the code should still continue because the parameters seem "good" enough.
```

**Review by wxiaoguang — APPROVED** [review]

**Review by lunny — APPROVED** [review]

**Pull request merged**